packages/@glimmer/runtime/lib/compiled/opcodes/dom.ts

```diff
@@ -93,16 +105,71 @@ APPEND_OPCODES.add(Op.Modifier, (vm, { op1: handle }) => {
     return;
   }
 
+  let owner = vm.getOwner();
+  let args = check(vm.stack.popJs(), CheckArguments);
   let definition = vm[CONSTANTS].getValue<ModifierDefinition>(handle);
-  let { manager } = definition;
 
-  let stack = vm.stack;
+  invokeModifier(vm, owner, args, definition);
+});
+
+APPEND_OPCODES.add(Op.DynamicModifier, (vm) => {
+  if (vm.env.isInteractive === false) {
+    return;
+  }
```

Comment:
It seems weird to bake this assumption into the VM layer, no?

Reply:
Looks like that ship has sailed unfortunately. 

https://github.com/glimmerjs/glimmer-vm/blob/3f29812a6b2904d2949a56c178babe1d9d8fa387/packages/%40glimmer/runtime/lib/compiled/opcodes/dom.ts#L92

I think this is conceptually wrong, and we should rethink but it is not a blocking concern for now.

Reply:
I think for the moment it doesn't really, it's pretty baked in already that modifiers should only run in the browser. If we want to change that, I think it's going to be a larger refactor necessarily, and probably will look like us checking the capability/timing of the manager + `vm.env.isInteractive`, but doing it in approximately the same spot.

If we don't do it here, then we end up creating, scheduling, and running everything for the modifier, even though it isn't actually going to do anything, which is pretty negative for perf overall.

Reply:
Its a bit odd to say its "very ingrained" since it is a new development in the VM space.

For example, it isn't baked into the VM of any release of Ember (all released versions of Ember that support modifiers, guard the interactivity in the Ember-side modifier manager).

Reply:
That's not exactly true, that code was a double guard. We've guarded modifiers actually running in the VM in the environment for some time: https://github.com/glimmerjs/glimmer-vm/blob/v0.56.0/packages/@glimmer/runtime/lib/environment.ts#L330 

But yeah, happy to revisit.

Reply:
For example, this change (currently slated for Ember 3.25 I think?) is plausibly introducing a breaking change for Ember users that use `data-ember-action` for styling **and** use FastBoot/SSR (this is a non-zero subset of users).

https://github.com/emberjs/ember.js/blob/907494ec67e516b7d213970545cd2527fe73001a/packages/%40ember/-internals/glimmer/lib/modifiers/action.ts#L280-L281

Reply:
Prior to that, it was guarded in the environment in Ember: https://github.com/emberjs/ember.js/blob/v3.16.10/packages/@ember/-internals/glimmer/lib/environment.ts#L83

So as far as I can tell, modifiers have never run in the browser, at least back to 3.16